[AutoFill Debugging] Add the ability to redact or replace the contents of AutoFilled text fields

whsieh opened the pull request:
#### 3fc2c409c17ec3bd8743ddcb6df38e3f160ed417
<pre>
[AutoFill Debugging] Add the ability to redact or replace the contents of AutoFilled text fields
<a href="https://bugs.webkit.org/show_bug.cgi?id=301956">https://bugs.webkit.org/show_bug.cgi?id=301956</a>
<a href="https://rdar.apple.com/163650051">rdar://163650051</a>

Reviewed by Abrar Rahman Protyasha and Aditya Keerthi.

Add support for a couple new properties, which allow clients to:

- Automatically replace all occurrences of any given set of strings with another set of strings.
- Avoid extracting text from all text inputs that have been modified via AutoFill.

Tests: fast/text-extraction/debug-text-extraction-ignore-autofilled-fields.html
       TextExtractionTests.ReplacementStrings

Tests: fast/text-extraction/debug-text-extraction-ignore-autofilled-fields.html
       Tools/TestWebKitAPI/Tests/WebKitCocoa/TextExtractionTests.mm
* LayoutTests/fast/text-extraction/debug-text-extraction-basic.html:
* LayoutTests/fast/text-extraction/debug-text-extraction-highlight-text.html:
* LayoutTests/fast/text-extraction/debug-text-extraction-ignore-autofilled-fields-expected.txt: Added.
* LayoutTests/fast/text-extraction/debug-text-extraction-ignore-autofilled-fields.html: Added.
* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::hasEnclosingAutoFilledInput):
(WebCore::TextExtraction::collectText):
(WebCore::TextExtraction::extractItem):

Add plumbing to avoid extracting text from text nodes inside input elements that have been
AutoFilled.

* Source/WebCore/page/text-extraction/TextExtractionTypes.h:
* Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
(WebKit::TextExtractionAggregator::applyReplacements):

Honor the new `replacementStrings` dictionary here, by swapping out all replacement text after
filtering. Importantly, this happens after filtering so that the results of OCR-based filtering
don&apos;t result in false positives due to replacement text being significantly different from the DOM
text.

(WebKit::addPartsForText):
* Source/WebKit/Shared/TextExtractionToStringConversion.h:
(WebKit::TextExtractionOptions::TextExtractionOptions):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(extractReplacementStrings):
(-[WKWebView _debugTextWithConfiguration:completionHandler:]):
(-[WKWebView _requestTextExtractionInternal:completion:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
(-[_WKTextExtractionConfiguration init]):
(-[_WKTextExtractionConfiguration replacementStrings]):
(-[_WKTextExtractionConfiguration setReplacementStrings:]):

Add the new properties here.

* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
* Tools/TestRunnerShared/UIScriptContext/UIScriptControllerShared.cpp:
(WTR::toTextExtractionTestOptions):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextExtractionTests.mm:
(TestWebKitAPI::TEST(TextExtractionTests, ReplacementStrings)):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::createTextExtractionConfiguration):

Canonical link: <a href="https://commits.webkit.org/302561@main">https://commits.webkit.org/302561@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7de438c6deefd2549e92f5d624af09abe4800358

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129488 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1745 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40327 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136866 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80911 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9e0af0f8-5052-4198-bac6-a5bf962443fc) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131359 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1677 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1621 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98628 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66498 "Passed tests") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8d3e23c9-1998-4f16-8483-a2aa283aa1c2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132435 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1299 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115979 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79272 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/77a6906f-1296-441c-8867-5ecd9a42e4ba) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1220 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34110 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80142 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/109689 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34607 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139340 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1535 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1471 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/107153 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1577 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112320 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106988 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27244 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1251 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30838 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54225 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1606 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64969 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1426 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1460 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1528 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/fast/text-extraction/debug-text-extraction-ignore-autofilled-fields.html:
```diff
@@ -0,0 +1,52 @@
+</head>
+<body>
+<div><input type="email" placeholder="Enter your email" /></div>
+<div><input type="password" placeholder="Enter your password" /></div>
```
Nit, since it stands out compared to the number input below 😅 
```suggestion
<div><input type="email" placeholder="Enter your email"></div>
<div><input type="password" placeholder="Enter your password"></div>
```

whsieh replied:
Whoops! Fixed.

aprotyas commented on LayoutTests/fast/text-extraction/debug-text-extraction-lightweight.html:
```diff
@@ -91,6 +91,7 @@ <h4>Related Links</h4>
         includeNodeIdentifiers: false,
         includeEventListeners: false,
         includeAccessibilityAttributes: false,
+        includeTextInAutoFilledControls: true,
```
this doesn't need setting right?

whsieh replied:
Indeed — I'll remove it here.

aprotyas commented on Source/WebCore/page/text-extraction/TextExtraction.cpp:
```diff
@@ -92,11 +92,22 @@ namespace TextExtraction {
+static bool hasEnclosingAutoFilledInput(Node& node)
+{
+    RefPtr input = dynamicDowncast<HTMLInputElement>(node.shadowHost());
+    if (!input)
+        return false;
+
+    return input->autofilled() || input->autofilledAndViewable() || input->autofilledAndObscured();
```
nit?
```suggestion
    return input && (input->autofilled() || input->autofilledAndViewable() || input->autofilledAndObscured());
```

aprotyas replied:
meh, maybe readability suffers

whsieh replied:
Ah, yeah — I originally had that on one line, but then split it out to a separate early return for readability. I think I'll keep it as-is.

whsieh committed:
[AutoFill Debugging] Add the ability to redact or replace the contents of AutoFilled text fields
https://bugs.webkit.org/show_bug.cgi?id=301956
rdar://163650051

Reviewed by Abrar Rahman Protyasha and Aditya Keerthi.

Add support for a couple new properties, which allow clients to:

- Automatically replace all occurrences of any given set of strings with another set of strings.
- Avoid extracting text from all text inputs that have been modified via AutoFill.

Tests: fast/text-extraction/debug-text-extraction-ignore-autofilled-fields.html
       TextExtractionTests.ReplacementStrings

Tests: fast/text-extraction/debug-text-extraction-ignore-autofilled-fields.html
       Tools/TestWebKitAPI/Tests/WebKitCocoa/TextExtractionTests.mm
* LayoutTests/fast/text-extraction/debug-text-extraction-basic.html:
* LayoutTests/fast/text-extraction/debug-text-extraction-highlight-text.html:
* LayoutTests/fast/text-extraction/debug-text-extraction-ignore-autofilled-fields-expected.txt: Added.
* LayoutTests/fast/text-extraction/debug-text-extraction-ignore-autofilled-fields.html: Added.
* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::hasEnclosingAutoFilledInput):
(WebCore::TextExtraction::collectText):
(WebCore::TextExtraction::extractItem):

Add plumbing to avoid extracting text from text nodes inside input elements that have been
AutoFilled.

* Source/WebCore/page/text-extraction/TextExtractionTypes.h:
* Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
(WebKit::TextExtractionAggregator::applyReplacements):

Honor the new `replacementStrings` dictionary here, by swapping out all replacement text after
filtering. Importantly, this happens after filtering so that the results of OCR-based filtering
don't result in false positives due to replacement text being significantly different from the DOM
text.

(WebKit::addPartsForText):
* Source/WebKit/Shared/TextExtractionToStringConversion.h:
(WebKit::TextExtractionOptions::TextExtractionOptions):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(extractReplacementStrings):
(-[WKWebView _debugTextWithConfiguration:completionHandler:]):
(-[WKWebView _requestTextExtractionInternal:completion:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
(-[_WKTextExtractionConfiguration init]):
(-[_WKTextExtractionConfiguration replacementStrings]):
(-[_WKTextExtractionConfiguration setReplacementStrings:]):

Add the new properties here.

* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
* Tools/TestRunnerShared/UIScriptContext/UIScriptControllerShared.cpp:
(WTR::toTextExtractionTestOptions):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextExtractionTests.mm:
(TestWebKitAPI::TEST(TextExtractionTests, ReplacementStrings)):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::createTextExtractionConfiguration):

Canonical link: https://commits.webkit.org/302561@main
